### PR TITLE
Added error explanations for E0308, E0309, and E0310

### PR DESCRIPTION
**Pull request by cactorium**
For https://github.com/rust-lang/rust/issues/24407

### CODE REVIEW

File: src/librustc/diagnostics.rs [hunk 1]
```diff
@@ -303,10 +303,63 @@ number cannot be negative.
...
 a compile-time constant.
+"##, 
+
+E0308: r##"
+This error occurs when the compiler was unable to infer the concrete type of a
+variable. This error can occur for several cases, the most common of which is
+that there is a mismatch in the expected type that the compiler inferred, and
+the actual type that the user defined a variable as.
+
+let a: char = 7;    // An integral type can't contained in a character, so
```
Comment (apasel422) [inline src/librustc/diagnostics.rs]: This should be "can't _be_ contained".

Comment (cactorium) [inline src/librustc/diagnostics.rs]: Ah, sorry, fixed

File: src/librustc/diagnostics.rs [hunk 2]
```diff
@@ -303,10 +303,63 @@ number cannot be negative.
 E0307: r##"
 The length of an array is part of its type. For this reason, this length must be
 a compile-time constant.
+"##,
+
+E0308: r##"
+This error occurs when the compiler was unable to infer the concrete type of a
+variable. This error can occur for several cases, the most common of which is
+that there is a mismatch in the expected type that the compiler inferred, and
+the actual type that the user defined a variable as.
```
Comment (tomprogrammer) [inline src/librustc/diagnostics.rs]: You mean "…variable has." (forgot a 'h')

**Comment by cactorium** [comment]
Nah, it's correct as is. It's the weird part of English where you have to put preposition after the object of the verb. (eg. "Where are you going to?")

File: src/librustc/diagnostics.rs (hunk 2 above)
Comment (pnkfelix) [inline src/librustc/diagnostics.rs]: I agree with @cactorium  that original is grammatically correct but also difficult to read.

I suggest perhaps:
> ... the most common of which is a mismatch in the expected type that the compiler inferred for a variable's initializing expression, and the actual type explicitly assigned to the variable.
> 
> For example:
> ```rust
> let x: i32 = "I am not a number!";
> //     ~~~   ~~~~~~~~~~~~~~~~~~~~
> //      |             |
> //      |    initializing expression;
> //      |    compiler infers type `&str`
> //      |
> //    type `i32` assigned to variable `x`
> ```


(Or drop the example if you're not into that. :)  )

Comment (cactorium) [inline src/librustc/diagnostics.rs]: That's way better worded :)! Thanks!

**Pull request merged**